### ISSUE TITLE
[Bug]: Unquantized BatchedTritonExperts can change greedy top-1 across identical runs

### ISSUE THREAD

**Issue by bossjoker1** [vllm-project/vllm#53430]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
OS: Ubuntu 22.04.5 LTS
Kernel: Linux 5.15.0-173-generic
Machine: x86_64
Python: 3.10.12
vLLM: 0.27.1+cu129
PyTorch: 2.13.0+cu130
CUDA used to build PyTorch: 13.0
Triton: 3.7.1
Transformers: 5.15.1
GPU: NVIDIA GeForce RTX 4090
Device capability: 8.9
Driver: 580.126.09
GPU memory: 49140 MiB
```

</details>

## Summary

On vLLM `0.27.1`, repeated calls to the unquantized `BatchedTritonExperts` path can return different BF16 tensors for exactly the same inputs. The direct `fused_experts` and modular Triton controls are bitwise stable under the same check.

In a full 24-sparse-layer replay of a pinned public Qwen MoE model, the batched path's numerical variation changes the greedy next token across identical runs. The behavior is consistent with the vectorized CUDA `index_add_` in `TopKWeightAndReduceNaiveBatched.apply`. Temporarily restoring the previous per-expert reduction loop removes the repeat-run variation in my bounded check.

## Minimal reproduction

This reducer-only reproduction uses positive, row-normalized router weights and does not download a model:

```python
import torch
from vllm.model_executor.layers.fused_moe.topk_weight_and_reduce import (
    TopKWeightAndReduceNaiveBatched,
)

torch.manual_seed(1)
num_experts, num_tokens, hidden_size, topk = 60, 64, 1408, 4
device = "cuda"

topk_ids = torch.stack(
    [
        (torch.arange(num_tokens, device=device) + slot * 13) % num_experts
        for slot in range(topk)
    ],
    dim=1,
).to(torch.int32)
topk_weights = torch.rand(num_tokens, topk, device=device)
topk_weights /= topk_weights.sum(dim=1, keepdim=True)

counts = torch.bincount(topk_ids.flatten().long(), minlength=num_experts)
expert_output = torch.randn(
    num_experts,
    int(counts.max()),
    hidden_size,
    device=device,
    dtype=torch.bfloat16,
)

reducer = TopKWeightAndReduceNaiveBatched(rank=0)
outputs = [
    reducer.apply(None, expert_output, topk_weights, topk_ids, False).clone()
    for _ in range(20)
]
torch.cuda.synchronize()

for repeat, output in enumerate(outputs[1:], start=1):
    difference = (output.float() - outputs[0].float()).abs()
    print(
        repeat,
        torch.equal(output, outputs[0]),
        difference.max().item(),
        int(torch.count_nonzero(difference).item()),
    )
```

On my machine all 19 comparisons print `False`; the maximum absolute difference is `0.015625`, with approximately 11,000-12,300 changed output elements per call.

## Model-impact reproduction

I also attached `repro_vllm_unquantized_moe_path_split.py`. It has no dependency on my test harness and uses only PyTorch, Transformers, and vLLM:

```bash
python repro_vllm_unquantized_moe_path_split.py --repeats 10 > repro.json
```

This script downloads and pins:

```text
model: Qwen/Qwen1.5-MoE-A2.7B-Chat
revision: ec052fda178e241c7c443468d2fa1db6618996be
dtype: bfloat16
prompt: Continue the Python code with exactly one token:\n```python\ndef
```

It performs two checks:

1. It captures the input to the first sparse layer once, derives the router weights/IDs and expert tensors once, and invokes each vLLM MoE path repeatedly with those exact tensors.
2. It replays all 24 sparse layers in the same pinned Transformers model, changing only the sparse expert implementation, and records the final vocabulary logits and greedy next token.

The `batched_triton_loop_finalize` path is a diagnostic monkeypatch contained in the script. It replaces only `TopKWeightAndReduceNaiveBatched.apply` with the [per-expert reduction loop from vLLM
0.26.0](https://github.com/vllm-project/vllm/blob/v0.26.0/vllm/model_executor/layers/fused_moe/topk_weight_and_reduce.py).

### Fixed first-layer inputs

For 10 calls per path, comparing calls 2-10 with call 1:

| Path | Bitwise-equal repeats | Largest observed absolute difference |
| --- | ---: | ---: |
| direct `fused_experts` | 9 / 9 | 0 |
| modular Triton experts | 9 / 9 | 0 |
| batched Triton experts | 0 / 9 | 0.00390625 |

The inputs, weights, selected expert IDs, router weights, and output buffer shape are unchanged between these calls.

### Full model replay

For 10 identical forward passes:

| Sparse expert implementation | Greedy top-1 result |
| --- | --- |
| original Transformers blocks | token `73594` (three backticks), 10 / 10 |
| direct vLLM `fused_experts` | token `35597` (`calculate`), 10 / 10 |
| modular vLLM Triton experts | token `35597` (`calculate`), 10 / 10 |
| batched vLLM Triton experts | token `73594`, 5 / 10; token `35597`, 5 / 10 |

The captured router weights are model-generated rather than synthetic. Their row sums range from approximately `0.10` to `0.93` during this replay.

This is a model replay that directly invokes vLLM's MoE kernels; it is not a claim that the default single-GPU vLLM server selects `BatchedTritonExperts` for this model.

## Localization check

The current
[`TopKWeightAndReduceNaiveBatched.apply`](https://github.com/vllm-project/vllm/blob/v0.27.1/vllm/model_executor/layers/fused_moe/topk_weight_and_reduce.py)
gathers all routed contributions and finishes with:

```python
output.index_add_(0, t_idx, gathered.to(output.dtype))
```

[PyTorch documents that CUDA `index_add_` may behave nondeterministically](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.index_add_.html). The observed repeat-run variation is consistent with concurrent additions to the same token row.

As a diagnostic only, I restored the earlier per-expert loop locally. With three repeats:

- the fixed first-layer output was bitwise equal on both comparisons to the first call (`2/2`), while the unmodified batched path differed on both (`0/2`);
- the full replay produced the same greedy token on all three runs (`3/3`), while the unmodified batched path produced both candidate tokens.

This local result narrows the source neighborhood, but I have not benchmarked the old loop and am not proposing it as the final implementation.

## Expected behavior

Repeated calls to the same unquantized batched MoE path with identical inputs should produce stable outputs. In particular, an internal reduction-order race should not make a downstream greedy argmax alternate between tokens.

Bitwise agreement between different MoE kernel families is not required for this report. The failure is repeat-run nondeterminism within one path.

## Scope

- Reproduced on one RTX 4090 with unquantized BF16 inputs and vLLM `0.27.1`.
- Reproduced by explicitly constructing `BatchedTritonExperts`; no distributed EP,
  all-to-all transport, quantization, or live vLLM server is required.
- I have not measured prevalence across models, GPUs, dtypes, or batch sizes.
- I have not measured the performance cost of a deterministic replacement.

## Related issue

[`#52525`](https://github.com/vllm-project/vllm/issues/52525) reports ordering-dependent nondeterminism in a quantized Marlin path. This report is in an unquantized batched finalize path and reproduces nondeterminism across repeated calls with identical inputs.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by wuhangxian** [comment]
Thanks for the detailed report and the minimal reproducer — it made this very easy to reproduce and diagnose.

I opened #53440 to fix this. The root cause is that `index_add_` on CUDA uses non-deterministic atomic adds when multiple experts route to the same token. The fix replaces it with a scatter into a `(T, topk, K)` buffer (each position written at most once) followed by `ops.moe_sum` (the same deterministic reduction kernel already used by `TopKWeightAndReduceContiguous`).

Verified on RTX PRO 5000 (SM120): 19/19 repeats now bitwise-identical, plus edge-case and end-to-end `BatchedTritonExperts` tests all pass.

I also see #53442 from @taking-lying-flat with the same core approach — happy to defer to maintainers on which to merge.
